src/js/components/DiceValue.jsx

```diff
@@ -0,0 +1,159 @@
+import React from 'react';
+
+export default class DiceValue extends React.Component {
+  constructor(props) {
+    super(props);
+
+    /*
+    props = {
+      onRoll: function that accepts a value and roll results
+      onComponentsChanged: function that reports changes to the dice components/values to encapsulating components
+    }
+    */
+    const compos = this.props.components || [{multiplier: 1, d: 6}];
+
+    this.state = {
+      components: compos,
+      lastRolledValue: 0,
+      lastRolledText: "",
+    };
+  }
+
+  componentDidUpdate(prevProps, prevState, snapshot) {
+    if (this.state.components !== prevState.components && this.props.onComponentsChanged) {
+      this.props.onComponentsChanged(this.state.components);
+    }
+  }
+
+  rollDie(dieSides) {
+    return 1 + Math.floor(Math.random() * (dieSides - 1));
+  }
+
+  rollComponent(comp) {
+    if (comp.d === 1) {
+      return {
+        value: comp.multiplier,
+        text: comp.multiplier
+      };
+    }
+
+    let total = 0;
+    let text = '';
+    for (let i = 0; i < comp.multiplier; i++) {
+      const singleResult = this.rollDie(comp.d);
+      if (i > 0) {
+        text += ' + ';
+      }
+      text += `${singleResult}/${comp.d}`;
+      total += singleResult;
+    }
+
+    return {
+      value: total,
+      text
+    };
+  }
+
+  handleRollClicked() {
+    const allRollResults = this.state.components.map(each => this.rollComponent(each));
+    const rollResult = allRollResults.map(each => each.value).reduce((a, b) => a + b);
+    const rollText = allRollResults.map(each => each.text).filter(each => each !== '').join(' + ');
+
+    this.state.lastRolledValue = rollResult;
+    this.state.lastRolledText = rollText
```

Comment:
Never set the state directly, instead use `this.setState({ lastRolledValue: rollResult })`

https://reactjs.org/docs/react-component.html#setstate
https://reactjs.org/docs/react-component.html#state